src/dsp_tools/commands/xml_validate/reformat_input.py

```diff
@@ -0,0 +1,201 @@
+        root: root of an xml with qnames and comments removed
+
+    Returns:
+        Class instance with the information reformatted
+    """
+    shortcode = root.attrib["shortcode"]
+    default_ontology = root.attrib["default-ontology"]
+    resources = _deserialise_all_resources(root)
+    return DataDeserialised(shortcode=shortcode, default_onto=default_ontology, resources=resources)
+
+
+def _deserialise_all_resources(root: etree._Element) -> list[ResourceData]:
+    all_res = []
+    for res in root.iterchildren():
+        match res.tag:
+            case "resource":
+                all_res.append(_deserialise_one_resource(res))
+            case "annotation" | "region" | "link" | "video-segment" | "audio-segment":
+                all_res.append(_deserialise_dsp_base_resource(res))
+            case _:
+                pass
+    return all_res
+
+
+def _deserialise_dsp_base_resource(resource: etree._Element) -> ResourceData:
+    return ResourceData(
+        res_id=resource.attrib["id"],
+        res_class=resource.attrib["restype"],
+        label=resource.attrib["label"],
+        values=[],
+    )
+
+
+def _deserialise_one_resource(resource: etree._Element) -> ResourceData:
+    res_id = resource.attrib["id"]
+    values: list[ValueData] = []
+    for val in resource.iterchildren():
+        values.extend(_deserialise_one_property(val))
+    return ResourceData(
+        res_id=res_id,
+        res_class=resource.attrib["restype"],
+        label=resource.attrib["label"],
+        values=values,
+    )
+
+
+def _deserialise_one_property(prop_ele: etree._Element) -> list[ValueData]:  # noqa: PLR0911 (too-many-branches, return statements)
+    match prop_ele.tag:
+        case "boolean-prop":
+            return _deserialise_boolean_prop(prop_ele)
+        case "color-prop":
+            return _deserialise_color_prop(prop_ele)
+        case "date-prop":
+            return _deserialise_date_prop(prop_ele)
+        case "decimal-prop":
+            return _deserialise_decimal_prop(prop_ele)
+        case "geoname-prop":
+            return _deserialise_geoname_prop(prop_ele)
+        case "list-prop":
+            return _deserialise_list_prop(prop_ele)
+        case "integer-prop":
+            return _deserialise_int_prop(prop_ele)
+        case "resptr-prop":
+            return _deserialise_resptr_prop(prop_ele)
+        case "text-prop":
+            return _deserialise_text_prop(prop_ele)
+        case "time-prop":
+            return _deserialise_time_prop(prop_ele)
+        case "uri-prop":
+            return _deserialise_uri_prop(prop_ele)
+        case _:
+            return []
+
+
+def _deserialise_boolean_prop(prop: etree._Element) -> list[ValueData]:
+    prop_name = prop.attrib["name"]
+    all_vals: list[ValueData] = []
+    for val in prop.iterchildren():
+        txt = val.text if val.text is not None else ""
+        all_vals.append(BooleanValueData(prop_name=prop_name, prop_value=txt))
+    return all_vals
```

Comment:
*suggestion*: Reduce code duplication

There is a lot of code duplication here, most of the `_deserialize*` function only differ by the type they construct.

You could simplify this by extracting the duplicate code and pass in a function which constructs it, e.g.:

```python 
from typing import Callable

def _deserialise(prop: etree._Element, func: Callable[[str, str], ValueData]) -> list[ValueData]:
    prop_name = prop.attrib["name"]
    all_vals: list[ValueData] = []
    for val in prop.iterchildren():
        txt = val.text if val.text is not None else ""
        all_vals.append(func(prop_name=prop_name, prop_value=txt))
    return all_vals

def _deserialise_boolean_prop(prop: etree._Element) -> list[ValueData]:
    return _deserialise(prop, BooleanValueData)

def _deserialise_color_prop(prop: etree._Element) -> list[ValueData]:
    return _deserialise(prop, ColorValueData)

# and so on and so forth ...
```

You could then also inline `_deserialise(prop, BooleanValueData)` into the `match` statement above.

Reply:
That look really nice thanks for pointing it out, I never did something like that. I will try it out.

Reply:
I am too much influence by functional programming ;)

It is often quite easy to pass in a function if only a small portion of the code differs. Btw: I have tested out the code myself and it should work.

Reply:
It does, Updated it, now I just have to fight with mypy about the typing and then I'll merge it.